x/gov/abci.go

```diff
@@ -72,6 +72,85 @@ func EndBlocker(ctx sdk.Context, keeper *keeper.Keeper) error {
 		return err
 	}
 
+	// fetch active proposals that need to be checked for quorum
+	rng = collections.NewPrefixUntilPairRange[time.Time, uint64](ctx.BlockTime())
+	err = keeper.QuorumCheckQueue.Walk(ctx, rng, func(key collections.Pair[time.Time, uint64], quorumChecksDone uint64) (bool, error) {
+		proposal, err := keeper.Proposals.Get(ctx, key.K2())
+		if err != nil {
+			return false, err
+		}
+
+		params, err := keeper.Params.Get(ctx)
+		if err != nil {
+			return false, err
+		}
+
+		// remove from queue
+		err = keeper.QuorumCheckQueue.Remove(ctx, key)
+		if err != nil {
+			return false, err
+		}
+
+		if params.QuorumCheckCount == 0 {
+			return false, nil
+		}
+
+		proposalVotingPeriod := proposal.VotingEndTime.Sub(*proposal.VotingStartTime)
+		s := proposalVotingPeriod - *params.QuorumTimeout
+		// this subtraction should be 0 or positive, otherwise the quorum timout would be after the voting period ends
+		if s < 0 {
+			// if so we should not check for quorum
+			return false, nil
+		}
+
+		// check if proposal passed quorum
+		quorum, err := keeper.HasReachedQuorum(ctx, proposal)
+		if err != nil {
+			return false, err
+		}
+
+		if quorum {
+			if quorumChecksDone > 0 && !proposal.VotingStartTime.Add(*params.QuorumTimeout).After(ctx.BlockTime()) {
+				// proposal passed quorum after timeout, extend voting period.
+				// canonically, we consider the first quorum check (i.e. when quorumChecksDone is 0)
+				// to be "right after" the quorum timeout has elapsed, so if quorum is reached
+				// when quorumChecksDone is 0, we don't extend the voting period.
+				endTime := ctx.BlockTime().Add(*params.MaxVotingPeriodExtension)
+				if endTime.After(*proposal.VotingEndTime) {
+					proposal.VotingEndTime = &endTime
+					err = keeper.SetProposal(ctx, proposal)
+					if err != nil {
+						return false, err
+					}
+				}
+			}
+		} else if quorumChecksDone < params.QuorumCheckCount && proposal.VotingEndTime.After(ctx.BlockTime()) {
+			// proposal did not pass quorum and is still active, add back to queue with updated time key and counter.
+			// compute time interval between quorum checks
+			t := s / time.Duration(params.QuorumCheckCount)
+			// find time for next quorum check
+			nextQuorumCheckTime := key.K1().Add(t)
+			for !nextQuorumCheckTime.After(ctx.BlockTime()) {
+				// next quorum check time is in the past, so add enough time intervals to get to the next quorum check time in the future
+				nextQuorumCheckTime = nextQuorumCheckTime.Add(t)
+			}
```

Comment:
If we add `t` to the current `ctx.BlockTime()` instead of `key.K1()`, we don't need that loop. WDYT?

Reply:
It's something I considered, but would result in inconsistent checking intervals (would not increment exactly by `t` as they would be always shifted due to using `ctx.BlockTime()`). Besides, that loop should almost never be run, unless invervals end up being really short, and even if this happens, it would be executed only a few iterations I believe.

I guess the best thing to do would be to check these things - as everything else really - with tests that provide full code coverage

Reply:
We need to be 200% sure that `t` will be always strictly positive, or else we have infinite loop.

Reply:
Right. Well since params validation prevents from adding a `QuorumTimeout` > `VotingPeriod` this is the case. And is especially true now where we store a snapshot of the params value (or sort of) when we add the proposal to the `QuorumCheckQueue`.

Reply:
OK sounds safe, thanks for the additional info